File: Modique.Frontend/src/app/services/product.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

export interface ProductImage {
  productImageId: number;
  imageUrl: string;
  order: number;
  isMain: boolean;
}

export interface Product {
  productId: number;
  name: string;
  description?: string;
  price: number;
  createdAt: string;
  isActive: boolean;
  categoryId: number;
  categoryName?: string;
  brandId: number;
  brandName?: string;
  images?: ProductImage[];
  colorIds?: number[];
  sizeIds?: number[];
}

export interface PagedResult<T> {
  items: T[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
  hasPreviousPage: boolean;
  hasNextPage: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  private apiUrl = 'http://localhost:5097/api/products';

  constructor(private http: HttpClient) { }

  getProducts(page: number = 1, pageSize: number = 20): Observable<PagedResult<Product>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('pageSize', pageSize.toString());
    
    return this.http.get<PagedResult<Product>>(this.apiUrl, { params });
  }

  getProductsForAdmin(page: number = 1, pageSize: number = 100): Observable<PagedResult<Product>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('pageSize', pageSize.toString());
    
    return this.http.get<PagedResult<Product>>(`${this.apiUrl}/admin`, { params });
  }

  getProductById(id: number): Observable<Product> {
    return this.http.get<Product>(`${this.apiUrl}/${id}`);
  }

  searchProducts(query: string, page: number = 1, pageSize: number = 20): Observable<PagedResult<Product>> {
    const params = new HttpParams()
      .set('query', query)
      .set('page', page.toString())
      .set('pageSize', pageSize.toString());
    
    return this.http.get<PagedResult<Product>>(`${this.apiUrl}/search`, { params });
  }

  createProduct(product: CreateProductDto): Observable<Product> {
    return this.http.post<Product>(this.apiUrl, product);
  }

  updateProduct(id: number, product: UpdateProductDto): Observable<Product> {
    return this.http.put<Product>(`${this.apiUrl}/${id}`, product);
  }

  deleteProduct(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getCategories(): Observable<Category[]> {
    return this.http.get<Category[]>('http://localhost:5097/api/categories');
  }

  getBrands(): Observable<Brand[]> {
    return this.http.get<Brand[]>('http://localhost:5097/api/brands');
  }

  createBrand(brand: CreateBrandDto): Observable<Brand> {
    return this.http.post<Brand>('http://localhost:5097/api/brands', brand);
  }

  updateBrand(id: number, brand: UpdateBrandDto): Observable<Brand> {
    return this.http.put<Brand>(`http://localhost:5097/api/brands/${id}`, brand);
  }

  deleteBrand(id: number): Observable<void> {
    return this.http.delete<void>(`http://localhost:5097/api/brands/${id}`);
  }
}

export interface CreateProductDto {
  name: string;
  description?: string;
  price: number;
  categoryId: number;
  brandId: number;
  isActive?: boolean;
  imageUrls?: string[];
  colorIds?: number[];
  sizeIds?: number[];
}

export interface UpdateProductDto {
  name: string;
  description?: string;
  price: number;
  categoryId: number;
  brandId: number;
  isActive: boolean;
  imageUrls?: string[];
  colorIds?: number[];
  sizeIds?: number[];
}

export interface Category {
  categoryId: number;
  name: string;
  subCategory: string;
  description: string;
}

export interface Brand {
  brandId: number;
  name: string;
  country: string;
  logoURL: string;
}

export interface CreateBrandDto {
  name: string;
  country: string;
  logoURL: string;
}

export interface UpdateBrandDto {
  brandId: number;
  name: string;
  country: string;
  logoURL: string;
}

File: Modique.Backend/Modique.Application/Services/ProductService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Modique.Application.DTOs.Common;
using Modique.Application.DTOs.Products;
using Modique.Application.Interfaces;
using Modique.Domain.Entities;
using Modique.Infrastructure.Data;

namespace Modique.Application.Services;

public class ProductService : IProductService
{
    private readonly ModiqueDbContext _db;
    private readonly ILogger<ProductService> _logger;

    public ProductService(ModiqueDbContext db, ILogger<ProductService> logger)
    {
        _db = db;
        _logger = logger;
    }

    public async Task<PagedResult<ProductDto>> GetAllAsync(int page = 1, int pageSize = 20)
    {
        var query = _db.Products
            .Include(p => p.Category)
            .Include(p => p.Brand)
            .Include(p => p.Images)
            .Where(p => p.IsActive)
            .AsQueryable();

        var totalCount = await query.CountAsync();

        var products = await query
            .OrderBy(p => p.Name)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        var productDtos = products.Select(p => new ProductDto
        {
            ProductId = p.ProductId,
            Name = p.Name,
            Description = p.Description,
            Price = p.Price,
            CreatedAt = p.CreatedAt,
            IsActive = p.IsActive,
            CategoryId = p.CategoryId,
            CategoryName = p.Category?.Name,
            BrandId = p.BrandId,
            BrandName = p.Brand?.Name,
            Images = p.Images.OrderBy(i => i.Order).Select(i => new DTOs.Products.ProductImageDto
            {
                ProductImageId = i.ProductImageId,
                ImageUrl = i.ImageUrl,
                Order = i.Order,
                IsMain = i.IsMain
            }).ToList()
        }).ToList();

        return new PagedResult<ProductDto>
        {
            Items = productDtos,
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<PagedResult<ProductDto>> GetAllForAdminAsync(int page = 1, int pageSize = 20)
    {
        var query = _db.Products
            .Include(p => p.Category)
            .Include(p => p.Brand)
            .Include(p => p.Images)
            .AsQueryable();

        var totalCount = await query.CountAsync();

        var products = await query
            .OrderBy(p => p.Name)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToListAsync();

        var productDtos = products.Select(p => new ProductDto
        {
            ProductId = p.ProductId,
            Name = p.Name,
            Description = p.Description,
            Price = p.Price,
            CreatedAt = p.CreatedAt,
            IsActive = p.IsActive,
            CategoryId = p.CategoryId,
            CategoryName = p.Category?.Name,
            BrandId = p.BrandId,
            BrandName = p.Brand?.Name,
            Images = p.Images.OrderBy(i => i.Order).Select(i => new DTOs.Products.ProductImageDto
            {
                ProductImageId = i.ProductImageId,
                ImageUrl = i.ImageUrl,
                Order = i.Order,
                IsMain = i.IsMain
            }).ToList()
        }).ToList();

        return new PagedResult<ProductDto>
        {
            Items = productDtos,
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<ProductDto?> GetByIdAsync(int id)
    {
        var product = await _db.Products
            .Include(p => p.Category)
            .Include(p => p.Brand)
            .Include(p => p.Images)
            .FirstOrDefaultAsync(p => p.ProductId == id);

        if (product == null)
            return null;

        return new ProductDto
        {
            ProductId = product.ProductId,
            Name = product.Name,
            Description = product.Description,
            Price = product.Price,
            CreatedAt = product.CreatedAt,
            IsActive = product.IsActive,
            CategoryId = product.CategoryId,
            CategoryName = product.Category?.Name,
            BrandId = product.BrandId,
            BrandName = product.Brand?.Name,
            Images = product.Images.OrderBy(i => i.Order).Select(i => new DTOs.Products.ProductImageDto
            {
                ProductImageId = i.ProductImageId,
                ImageUrl = i.ImageUrl,
                Order = i.Order,
                IsMain = i.IsMain
            }).ToList()
        };
    }

    public async Task<ProductDto> CreateAsync(CreateProductDto dto)
    {
        var product = new Product
        {
            Name = dto.Name,
            Description = dto.Description,
            Price = dto.Price,
            CategoryId = dto.CategoryId,
            BrandId = dto.BrandId,
            IsActive = dto.IsActive,
            CreatedAt = DateTime.UtcNow
        };

        _db.Products.Add(product);
        await _db.SaveChangesAsync();

        if (dto.ImageUrls != null && dto.ImageUrls.Any())
        {
            for (int i = 0; i < dto.ImageUrls.Count; i++)
            {
                var image = new Domain.Entities.ProductImage
                {
                    ProductId = product.ProductId,
                    ImageUrl = dto.ImageUrls[i],
                    Order = i,
                    IsMain = i == 0
                };
                _db.ProductImages.Add(image);
            }
            await _db.SaveChangesAsync();
        }

        _logger.LogInformation("Product created: {ProductId} - {ProductName}", product.ProductId, product.Name);

        return await GetByIdAsync(product.ProductId) ?? throw new InvalidOperationException("Failed to retrieve created product");
    }

    public async Task<ProductDto?> UpdateAsync(int id, UpdateProductDto dto)
    {
        var product = await _db.Products
            .Include(p => p.Images)
            .FirstOrDefaultAsync(p => p.ProductId == id);
        if (product == null)
            return null;

        product.Name = dto.Name;
        product.Description = dto.Description;
        product.Price = dto.Price;
        product.CategoryId = dto.CategoryId;
        product.BrandId = dto.BrandId;
        product.IsActive = dto.IsActive;

        if (dto.ImageUrls != null && dto.ImageUrls.Any())
        {
            var existingImages = product.Images.ToList();
            _db.ProductImages.RemoveRange(existingImages);

            for (int i = 0; i < dto.ImageUrls.Count; i++)
            {
                var image = new Domain.Entities.ProductImage
                {
                    ProductId = product.ProductId,
                    ImageUrl = dto.ImageUrls[i],
                    Order = i,
                    IsMain = i == 0
                };
                _db.ProductImages.Add(image);
            }
        }

        await _db.SaveChangesAsync();

        _logger.LogInformation("Product updated: {ProductId}", product.ProductId);

        return await GetByIdAsync(id);
    }

    public async Task<bool> DeleteAsync(int id)
    {
        var product = await _db.Products.FindAsync(id);
        if (product == null)
            return false;

        _db.Products.Remove(product);
        await _db.SaveChangesAsync();

        _logger.LogInformation("Product deleted: {ProductId}", id);

        return true;
    }
}




